### PR TITLE
Avoid actually allocating a giant buffer

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

Our CI system is beginning to struggle with allocating a giant buffer in testSliceOfMassiveBufferWithAdvancedReaderIndexIsOk. We can work around this by faking out the allocation.

Modifications:

- Provide a fake allocator that doesn't actually allocate memory.
- Rewrite the test to use it.

Result:

Test still validates the behaviour but doesn't touch memory anymore.

### CODE REVIEW

File: Tests/NIOCoreTests/ByteBufferTest.swift [hunk 1]
```diff
@@ -1754,14 +1754,24 @@ class ByteBufferTest: XCTestCase {
             throw XCTSkip("This test is only supported on 64-bit systems.")
```
Comment (dnadoba) [inline Tests/NIOCoreTests/ByteBufferTest.swift:1754]: Can we now run this on 32-bit systems as well?

**Review by dnadoba — APPROVED** [review]
LGTM!

File: Tests/NIOCoreTests/ByteBufferTest.swift (hunk 1 above)
Comment (Lukasa) [reply Tests/NIOCoreTests/ByteBufferTest.swift:1754]: Unfortunately no: we still need to be able to have a buffer size that is 32-bits long, but we store the sizes in `Int`.

**Review by FranzBusch — APPROVED** [review]

File: Tests/NIOCoreTests/ByteBufferTest.swift [hunk 2]
```diff
@@ -1754,14 +1754,24 @@ class ByteBufferTest: XCTestCase {
...
+
         let targetSize = Int(UInt32.max)
-        var buffer = self.allocator.buffer(capacity: targetSize)
+        var buffer = fakeAllocator.buffer(capacity: targetSize)
 
         // Move the reader index forward such that we hit the slow path.
         let offset = Int(_UInt24.max) + 1
         buffer.moveWriterIndex(to: offset)
         buffer.moveReaderIndex(to: offset)
-        buffer.writeInteger(UInt32(0))
```
Comment (weissi) [inline Tests/NIOCoreTests/ByteBufferTest.swift:1764]: @Lukasa you can still make this write happen by having the fake allocator do

```
let ptr = malloc(4)
return ptr - offset
```

And `free` would do `free(ptr + offset)`.

Comment (Lukasa) [reply Tests/NIOCoreTests/ByteBufferTest.swift:1764]: Sure, but the write isn't important: we never validated the bytes came back.

File: Tests/NIOCoreTests/ByteBufferTest.swift (hunk 1 above)
Comment (buttaface) [reply Tests/NIOCoreTests/ByteBufferTest.swift:1754]: Unfortunately, the two `0xdeadbeef` lines you added below [broke compilation on 32-bit platforms like Android armv7 too](https://github.com/buttaface/swift-android-sdk/actions/runs/3909881597/jobs/6681462361#step:13:963), so this test will either need to be conditionally compiled only on 64-bit platforms or modified in some other way.

Comment (Lukasa) [reply Tests/NIOCoreTests/ByteBufferTest.swift:1754]: Presumably the easiest fix is to just use a different sentinel value. The number doesn't matter: do you want to open a quick PR to use a different number?

Comment (buttaface) [reply Tests/NIOCoreTests/ByteBufferTest.swift:1754]: As in change it to a 32-bit number that could be stored in an Int32 also? It's your idea, please go ahead.

Comment (Lukasa) [reply Tests/NIOCoreTests/ByteBufferTest.swift:1754]: Should be fixed by #2348.

Comment (buttaface) [reply Tests/NIOCoreTests/ByteBufferTest.swift:1754]: Yep, [compiles now](https://github.com/buttaface/swift-android-sdk/actions/runs/3917797023/jobs/6697815544), thanks.

**Pull request merged**